File: tests/java/com/pdfjet/TextBlockTest.java
/*
 * TextBlockTest.java
 *
 * Copyright (c) 2026 PDFjet Software
 * Licensed under the MIT License. See LICENSE file in the project root.
 */
package com.pdfjet;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertSame;
import static org.junit.jupiter.api.Assertions.assertTrue;

import org.junit.jupiter.api.Test;

class TextBlockTest {
    private static final String TEN_WORDS = "one two three four five six seven eight nine ten";

    @Test
    void aNewlineIsOneEmptyLine() throws Exception {
        Font font = TestSupport.helvetica(TestSupport.newPDF());
        float[] x = new TextBlock(font, "x").setLocation(0f, 0f).drawOn(null);
        TestSupport.assertXY(500f, 13.872f, x);
        TestSupport.assertXY(x[0], x[1], new TextBlock(font, "\n").setLocation(0f, 0f).drawOn(null));
        TestSupport.assertXY(x[0], x[1], new TextBlock(font, "").setLocation(0f, 0f).drawOn(null));
    }

    @Test
    void withoutAHeightTheBlockIsAsTallAsItsText() throws Exception {
        Font font = TestSupport.helvetica(TestSupport.newPDF());
        TextBlock block = new TextBlock(font, TEN_WORDS).setLocation(0f, 0f).setWidth(60f);
        // Six lines of 13.872 points.
        TestSupport.assertXY(60f, 83.232f, block.drawOn(null));
        assertEquals(83.232f, block.getHeight(), TestSupport.DELTA);
        assertEquals(60f, block.getWidth(), 0f);
    }

    @Test
    void aHeightCutsTheTextThatDoesNotFitAndAlignsTheRest() throws Exception {
        PDF pdf = TestSupport.newPDF();
        Font font = TestSupport.helvetica(pdf);
        TextBlock block = new TextBlock(font, TEN_WORDS).setLocation(0f, 0f).setSize(60f, 30f);
        // Two of the six lines fit
        TestSupport.assertXY(60f, 30f, block.drawOn(null));
        assertEquals(30f, block.getHeight(), 0f);
        Page page = new Page(pdf, Letter.PORTRAIT);
        block.drawOn(page);
        String content = TestSupport.content(page);
        assertTrue(content.contains(TestSupport.hex("...")), content);
        assertFalse(content.contains(TestSupport.hex("ten")), content);

        // Aligned to the bottom of a block 10 points taller than its two lines,
        // the text sits where a block without a height draws it 10 points lower
        page = new Page(pdf, Letter.PORTRAIT);
        block.setSize(60f, 2 * 13.872f + 10f).setVerticalAlignment(Alignment.BOTTOM).drawOn(page);
        String bottom = TestSupport.content(page);
        page = new Page(pdf, Letter.PORTRAIT);
        new TextBlock(font, "one two three four").setLocation(0f, 10f).setWidth(60f).drawOn(page);
        String lower = TestSupport.content(page);
        String textMatrix = lower.substring(lower.indexOf("1 0 0 1 0 "), lower.indexOf(" Tm\n") + 4);
        assertTrue(bottom.contains(textMatrix), textMatrix + " missing from " + bottom);
    }

    @Test
    void strikeoutDrawsALineThroughEachLine() throws Exception {
        PDF pdf = TestSupport.newPDF();
        Font font = TestSupport.helvetica(pdf);
        Page page = new Page(pdf, Letter.PORTRAIT);
        new TextBlock(font, "one\ntwo").setLocation(0f, 0f).setStrikeout(true).drawOn(page);
        String content = TestSupport.content(page);
        assertEquals(2, content.split(" l\n").length - 1, content);
    }

    @Test
    void drawOnAPageWritesEveryWord() throws Exception {
        PDF pdf = TestSupport.newPDF();
        Page page = new Page(pdf, Letter.PORTRAIT);
        float[] xy = new TextBlock(TestSupport.helvetica(pdf), TEN_WORDS).setLocation(0f, 0f).setWidth(60f).drawOn(page);
        TestSupport.assertXY(60f, 83.232f, xy);
        String content = TestSupport.content(page);
        assertTrue(content.contains(TestSupport.hex("one")), content);
        assertTrue(content.contains(TestSupport.hex("ten")), content);
    }

    @Test
    void transparentLeavesTheTextColorUnchanged() throws Exception {
        TextBlock block = new TextBlock(TestSupport.helvetica(TestSupport.newPDF()), "x");
        block.setTextColor(Color.blue).setTextColor(Color.transparent);
        TestSupport.assertRGB(0f, 0f, 1f, block.getTextColor());
    }

    @Test
    void setFontChangesTheFallbackFontUnlessAnotherWasSet() throws Exception {
        PDF pdf = TestSupport.newPDF();
        Font helvetica = TestSupport.helvetica(pdf);
        Font courier = new Font(pdf, CoreFont.COURIER);
        TextBlock block = new TextBlock(helvetica, "x").setFont(courier);
        assertSame(courier, block.fallbackFont);
        block.setFallbackFont(helvetica).setFont(new Font(pdf, CoreFont.TIMES_ROMAN));
        assertSame(helvetica, block.fallbackFont);
    }
}

File: tests/dotnet/TextBlockTest.cs
/*
 * TextBlockTest.cs
 *
 * Copyright (c) 2026 PDFjet Software
 * Licensed under the MIT License. See LICENSE file in the project root.
 */
using Xunit;

namespace PDFjet.NET {
public class TextBlockTest {
    private const string TEN_WORDS = "one two three four five six seven eight nine ten";

    [Fact]
    public void ANewlineIsOneEmptyLine() {
        Font font = TestSupport.Helvetica(TestSupport.NewPDF());
        float[] x = new TextBlock(font, "x").SetLocation(0f, 0f).DrawOn(null);
        TestSupport.AssertXY(500f, 13.872f, x);
        TestSupport.AssertXY(x[0], x[1], new TextBlock(font, "\n").SetLocation(0f, 0f).DrawOn(null));
        TestSupport.AssertXY(x[0], x[1], new TextBlock(font, "").SetLocation(0f, 0f).DrawOn(null));
    }

    [Fact]
    public void WithoutAHeightTheBlockIsAsTallAsItsText() {
        Font font = TestSupport.Helvetica(TestSupport.NewPDF());
        TextBlock block = new TextBlock(font, TEN_WORDS).SetLocation(0f, 0f).SetWidth(60f);
        // Six lines of 13.872 points.
        TestSupport.AssertXY(60f, 83.232f, block.DrawOn(null));
        TestSupport.AssertNear(83.232f, block.GetHeight(), TestSupport.DELTA);
        Assert.Equal(60f, block.GetWidth());
    }

    [Fact]
    public void AHeightCutsTheTextThatDoesNotFitAndAlignsTheRest() {
        PDF pdf = TestSupport.NewPDF();
        Font font = TestSupport.Helvetica(pdf);
        TextBlock block = new TextBlock(font, TEN_WORDS).SetLocation(0f, 0f).SetSize(60f, 30f);
        // Two of the six lines fit
        TestSupport.AssertXY(60f, 30f, block.DrawOn(null));
        Assert.Equal(30f, block.GetHeight());
        Page page = new Page(pdf, Letter.PORTRAIT);
        block.DrawOn(page);
        string content = TestSupport.Content(page);
        Assert.Contains(TestSupport.Hex("..."), content);
        Assert.DoesNotContain(TestSupport.Hex("ten"), content);

        // Aligned to the bottom of a block 10 points taller than its two lines,
        // the text sits where a block without a height draws it 10 points lower
        page = new Page(pdf, Letter.PORTRAIT);
        block.SetSize(60f, 2 * 13.872f + 10f).SetVerticalAlignment(Alignment.BOTTOM).DrawOn(page);
        string bottom = TestSupport.Content(page);
        page = new Page(pdf, Letter.PORTRAIT);
        new TextBlock(font, "one two three four").SetLocation(0f, 10f).SetWidth(60f).DrawOn(page);
        string lower = TestSupport.Content(page);
        int start = lower.IndexOf("1 0 0 1 0 ");
        string textMatrix = lower.Substring(start, lower.IndexOf(" Tm\n") + 4 - start);
        Assert.Contains(textMatrix, bottom);
    }

    [Fact]
    public void PaddingWiderThanTheBlockDrawsOneCharacterALine() {
        PDF pdf = TestSupport.NewPDF();
        Font font = TestSupport.Helvetica(pdf);
        Page page = new Page(pdf, Letter.PORTRAIT);
        // The text area was negative, and the word was broken past its end.
        TextBlock block = new TextBlock(font, "Hello");
        block.SetLocation(50f, 50f).SetWidth(100f).SetPadding(60f).DrawOn(page);
        string content = TestSupport.Content(page);
        Assert.Contains(TestSupport.Hex("H"), content);
        Assert.Contains(TestSupport.Hex("o"), content);
    }

    [Fact]
    public void TheUnderlineIsDrawnInTheTextColorAtTheFontThickness() {
        PDF pdf = TestSupport.NewPDF();
        Font font = TestSupport.Helvetica(pdf);
        Page page = new Page(pdf, Letter.PORTRAIT);
        page.SetPenColor(Color.red);     // the underline took the pen color
        TextBlock block = new TextBlock(font, "Hello");
        block.SetLocation(50f, 50f).SetTextColor(Color.blue);
        block.SetUnderline(true);
        block.SetBorderWidth(3f);        // and the border width
        block.DrawOn(page);
        string content = TestSupport.Content(page);
        Assert.Contains("0 0 1 RG", content);
        Assert.DoesNotContain("3 w", content);
        Assert.Contains(font.GetUnderlineThickness(font.GetSize()) + " w", content);
    }

    [Fact]
    public void TheCornerRadiusRoundsABackgroundWithoutABorder() {
        PDF pdf = TestSupport.NewPDF();
        Font font = TestSupport.Helvetica(pdf);
        Page page = new Page(pdf, Letter.PORTRAIT);
        TextBlock block = new TextBlock(font, "Hello");
        block.SetLocation(50f, 50f).SetBackgroundColor(Color.yellow);
        block.SetCornerRadius(10f);
        block.DrawOn(page);
        // A rounded rectangle draws its corners with the curve operator.
        Assert.Contains(" c\n", TestSupport.Content(page));
    }

    [Fact]
    public void ANullTextColorLeavesTheTextColorUnchanged() {
        PDF pdf = TestSupport.NewPDF();
        Font font = TestSupport.Helvetica(pdf);
        Page page = new Page(pdf, Letter.PORTRAIT);
        page.SetBrushColor(Color.red);
        TextBlock block = new TextBlock(font, "Hi");
        block.SetLocation(50f, 50f);
        block.SetTextColor((float[]) null);
        block.DrawOn(page);
        TestSupport.AssertRGB(0f, 0f, 0f, block.GetTextColor());
        Assert.Contains("0 0 0 rg", TestSupport.Content(page));
    }

    [Fact]
    public void StrikeoutDrawsALineThroughEachLine() {
        PDF pdf = TestSupport.NewPDF();
        Font font = TestSupport.Helvetica(pdf);
        Page page = new Page(pdf, Letter.PORTRAIT);
        new TextBlock(font, "one\ntwo").SetLocation(0f, 0f).SetStrikeout(true).DrawOn(page);
        string content = TestSupport.Content(page);
        Assert.Equal(2, content.Split(" l\n").Length - 1);
    }

    [Fact]
    public void DrawOnAPageWritesEveryWord() {
        PDF pdf = TestSupport.NewPDF();
        Page page = new Page(pdf, Letter.PORTRAIT);
        float[] xy = new TextBlock(TestSupport.Helvetica(pdf), TEN_WORDS).SetLocation(0f, 0f).SetWidth(60f).DrawOn(page);
        TestSupport.AssertXY(60f, 83.232f, xy);
        string content = TestSupport.Content(page);
        Assert.Contains(TestSupport.Hex("one"), content);
        Assert.Contains(TestSupport.Hex("ten"), content);
    }

    [Fact]
    public void TransparentLeavesTheTextColorUnchanged() {
        TextBlock block = new TextBlock(TestSupport.Helvetica(TestSupport.NewPDF()), "x");
        block.SetTextColor(Color.blue).SetTextColor(Color.transparent);
        TestSupport.AssertRGB(0f, 0f, 1f, block.GetTextColor());
    }
}
}
